### PR TITLE
feat: Autoscaler

### PR DESCRIPTION
**Pull request by danieljbruce**
This feature allows the user to set properties when creating a cluster or updating a cluster that turn on the autoscaler or turn it off to use manual scaling.

### CODE REVIEW

File: src/cluster.ts [hunk 1]
```diff
@@ -75,7 +76,10 @@ export type GetClustersCallback = (
   apiResponse?: google.bigtable.admin.v2.IListClustersResponse
 ) => void;
 export interface SetClusterMetadataOptions {
-  nodes: number;
+  nodes?: number;
```
Comment (kolea2) [inline src/cluster.ts:79]: is this considered breaking if we change something from required to optional? I'm guessing it would be the other way around, but want to confirm :)

File: src/cluster.ts [hunk 2]
```diff
@@ -690,35 +697,42 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
...
+      );
+    };
+    if (this.metadata && this.metadata.location) {
+      setMetadataWithLocation(this.metadata.location, this.name);
+    } else {
+      this.getMetadata(gaxOptions, (err, res) => {
+        if (err || !this.metadata || !this.metadata.location) {
+          callback(err, null);
+        } else {
+          setMetadataWithLocation(this.metadata.location, this.name);
```
Comment (kolea2) [inline src/cluster.ts:732]: why do we need to reset the metadata in this case?

File: src/instance.ts [hunk 3]
```diff
@@ -415,17 +422,6 @@ Please use the format 'my-instance' or '${bigtable.projectName}/instances/my-ins
...
     }
 
-    if (options.location) {
-      reqOpts.cluster!.location = Cluster.getLocation_(
-        this.bigtable.projectId,
-        options.location
-      );
-    }
-
-    if (options.nodes) {
```
Comment (kolea2) [inline src/instance.ts:425]: why is this deleted?

File: src/cluster.ts (hunk 1 above)
Comment (danieljbruce) [reply src/cluster.ts:79]: Here we are changing something from required to optional so that would make it not a breaking change since users can still pass just `nodes` in. This is where we make `nodes` optional because instead the user can provide autoscaling parameters. These parameters will be validated against a new validation function and an error will be thrown if an invalid combination of parameters are provided.

File: src/instance.ts (hunk 3 above)
Comment (danieljbruce) [reply src/instance.ts:425]: `cluster.serveNodes` is now provided by the call to `ClusterUtils.getClusterBaseConfig` above. This change is a refactor so that this function doesn't get too verbose.

File: system-test/cluster.ts [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+          try {
+            await cluster2.create({
+              location: 'us-west1-c',
+            });
+            assert.fail();
+          } catch (e) {
+            assert.equal(e.message, ClusterUtils.noConfigError);
+          }
+        });
+        it('By providing too much cluster configurations', async () => {
```
Comment (kolea2) [inline system-test/cluster.ts:105]: nit - reword to express that the user entered manual AND autoscaling configurations

File: system-test/cluster.ts [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+          await localCluster.getMetadata({});
+        assert.strictEqual(metadata[0].serveNodes, nodes);
+      }
+      beforeEach(async () => {
+        clusterId = generateId('cluster');
+        await createStandardNewInstance(clusterId, 2);
+        cluster = instance.cluster(clusterId);
+      });
+      it('Create an instance with clusters for manual scaling', async () => {
+        await checkMetadata(cluster, 2);
```
Comment (kolea2) [inline system-test/cluster.ts:78]: should there be assertions here and below?

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: system-test/cluster.ts (hunk 5 above)
Comment (danieljbruce) [reply system-test/cluster.ts:78]: checkMetadata contains an assertion.

**Review by danieljbruce — COMMENTED** [review]
Initial follow-up to first review.

File: system-test/cluster.ts [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+      });
+      it('Create an instance and then create a cluster for manual scaling', async () => {
```
Comment (kolea2) [inline system-test/cluster.ts:80]: nit - here and in all other `it(..` descriptions, I think they should be reworded slightly; for example, here I would change it to say:

`it('should create an instance and then create a cluster with manual scaling')`

Let me know your thoughts :)

Comment (danieljbruce) [reply system-test/cluster.ts]: Very good point. I'll look at some of the other PRs and make sure that they have these descriptions as well.

File: src/utils/cluster.ts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  static getClusterBaseConfig(
+    metadata: SetClusterMetadataOptions | BasicClusterConfig,
+    location: string | undefined,
+    name: string | undefined
+  ): google.bigtable.admin.v2.ICluster {
+    let clusterConfig;
+    if (
+      metadata.cpuUtilizationPercent ||
+      metadata.minServeNodes ||
+      metadata.maxServeNodes
```
[the remark this answers is not in the document]
Comment (danieljbruce) [reply src/utils/cluster.ts:103]: When creating instances we now validate inputs.

**Review by danieljbruce — COMMENTED** [review]
Addressed two more outstanding comments

File: system-test/cluster.ts [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+              location: 'us-west1-c',
+              nodes: 2,
+              minServeNodes: 3,
+            });
+            assert.fail();
+          } catch (e) {
+            assert.equal(e.message, ClusterUtils.allConfigError);
+          }
+        });
+        it('should throw an error providing all autoscaling configurations', async () => {
```
Comment (kolea2) [inline system-test/cluster.ts:134]: should this say "should throw an error when missing some autoscaling configurations"?

File: system-test/cluster.ts [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+
+      beforeEach(async () => {
+        const clusterId = generateId('cluster');
+        await createStandardNewInstance(clusterId, startingNodes);
+        cluster = instance.cluster(clusterId);
+      });
+
+      it('should change nodes for manual scaling', async () => {
+        const updateNodes = 5;
+        assert.notEqual(startingNodes, updateNodes);
```
Comment (kolea2) [inline system-test/cluster.ts:204]: this check is comparing two int variables - if you want to have this, I would instead get the creater cluster's nodes, then compare the created cluster's nodes to the updateNodes value.

File: system-test/cluster.ts [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+            await cluster.setMetadata({
+              nodes: 2,
+              minServeNodes: 3,
+            });
+            assert.fail();
+          } catch (e) {
+            assert.equal(e.message, ClusterUtils.allConfigError);
+          }
+        });
+        it('should throw an error providing all autoscaling configurations', async () => {
```
Comment (kolea2) [inline system-test/cluster.ts:254]: i think this is the wrong description for this test?

**Review by kolea2 — CHANGES_REQUESTED** [review]

File: system-test/cluster.ts (hunk 9 above)
Comment (danieljbruce) [reply system-test/cluster.ts:204]: `checkMetadata` does the second thing you mentioned though it checks `metadata.serveNodes` equals `updateNodes` rather than checking `metadata.serveNodes` does not equal `startingNodes`. This `assert.notEqual` is more of a maintainability thing to ensure the programmer doesn't set `updateNodes` to equal `startingNodes` and then have the test pass when `setMetadata` doesn't actually work properly. If you think it's confusing though then we can remove it.

File: system-test/cluster.ts [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+            ...createClusterOptions,
+            nodes: minServeNodes,
+          },
+          true
+        );
+      });
+    });
+  });
+  describe('Update cluster', () => {
+    describe('Starting from manual scaling', () => {
```
[the remark this answers is not in the document]
Comment (danieljbruce) [reply system-test/cluster.ts:192]: Done

File: system-test/cluster.ts (hunk 10 above)
Comment (danieljbruce) [reply system-test/cluster.ts:254]: Yeah. This is a confusing error. Related to the comment above. Changing this now.

File: system-test/cluster.ts (hunk 9 above)
Comment (kolea2) [reply system-test/cluster.ts:204]: I don't think this is providing much in terms of autoscaling assertions, so I would remove it in favor of the checks you already mentioned.

File: src/cluster.ts [hunk 12]
```diff
@@ -690,35 +697,42 @@ Please use the format 'my-cluster' or '${instance.name}/clusters/my-cluster'.`);
...
-        client: 'BigtableInstanceAdminClient',
-        method: 'updateCluster',
-        reqOpts,
-        gaxOpts: gaxOptions,
-      },
-      (err, resp) => {
-        callback(err, resp);
-      }
-    );
+    const setMetadataWithLocation = (location: string, name: string) => {
```
Comment (bcoe) [inline src/cluster.ts:707]: I would remove this logic if we think it's currently not needed based on your testing.

File: src/utils/cluster.ts [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+    location: string,
+    name: string
+  ): google.bigtable.admin.v2.ICluster {
+    const cluster: ICluster = Object.assign(
+      {},
+      this.getClusterBaseConfig(metadata, location, name),
+      metadata
+    );
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    delete (cluster as any).nodes;
```
Comment (bcoe) [inline src/utils/cluster.ts:137]: I don't think you should have to case to `any`, assuming that these are properties on `ICluster`.

**Review by bcoe — CHANGES_REQUESTED** [review]

File: src/cluster.ts (hunk 1 above)
Comment (bcoe) [reply src/cluster.ts:79]: I would not consider switching from required to optional for TypeScript breaking 👍

**Review by bcoe — APPROVED** [review]

**Review by kolea2 — APPROVED** [review]
